src/arkode/arkode_user_controller.c

```diff
@@ -0,0 +1,183 @@
+#define SC_HP(C)          ( SC_CONTENT(C)->hp )
+#define SC_HPP(C)         ( SC_CONTENT(C)->hpp )
+#define SC_EP(C)          ( SC_CONTENT(C)->ep )
+#define SC_EPP(C)         ( SC_CONTENT(C)->epp )
+#define SC_P(C)           ( SC_CONTENT(C)->p )
+#define SC_Q(C)           ( SC_CONTENT(C)->q )
+#define SC_ARKMEM(C)      ( SC_CONTENT(C)->ark_mem )
+#define SC_HADAPT(C)      ( SC_CONTENT(C)->hadapt )
+#define SC_DATA(C)        ( SC_CONTENT(C)->hadapt_data )
+
+
+/* -----------------------------------------------------------------
+ * exported functions
+ * ----------------------------------------------------------------- */
+
+/* -----------------------------------------------------------------
+ * Function to create a new ARKUserControl controller
+ */
+
+SUNControl ARKUserControl(SUNContext sunctx, void* arkode_mem,
+                          ARKAdaptFn hadapt, void* hadapt_data)
+{
+  SUNControl C;
+  ARKUserControlContent content;
+
+  /* Return with failure if hadapt or arkode_mem are NULL */
+  if ((hadapt == NULL) || (arkode_mem == NULL)) { return (NULL); }
+
+  /* Create an empty controller object */
+  C = NULL;
+  C = SUNControlNewEmpty(sunctx);
+  if (C == NULL) { return (NULL); }
+
+  /* Attach operations */
+  C->ops->getid             = SUNControlGetID_ARKUserControl;
+  C->ops->estimatestep      = SUNControlEstimateStep_ARKUserControl;
+  C->ops->reset             = SUNControlReset_ARKUserControl;
+  C->ops->write             = SUNControlWrite_ARKUserControl;
+  C->ops->setmethodorder    = SUNControlSetMethodOrder_ARKUserControl;
+  C->ops->setembeddingorder = SUNControlSetEmbeddingOrder_ARKUserControl;
+  C->ops->update            = SUNControlUpdate_ARKUserControl;
+  C->ops->space             = SUNControlSpace_ARKUserControl;
+
+  /* Create content */
+  content = NULL;
+  content = (ARKUserControlContent)malloc(sizeof *content);
+  if (content == NULL)
+  {
+    SUNControlDestroy(C);
+    return (NULL);
+  }
+
+  /* Attach content */
+  C->content = content;
+
+  /* Attach ARKODE memory structure */
+  content->ark_mem = (ARKodeMem) arkode_mem;
+
+  /* Attach user-provided adaptivity function and data */
+  content->hadapt = hadapt;
+  content->hadapt_data = hadapt_data;
+
+  /* Initialize method and embedding orders */
+  content->p = 1;
+  content->q = 1;
+
+  /* Fill content with default/reset values */
+  SUNControlReset_ARKUserControl(C);
+
+  return (C);
+}
+
+
+/* -----------------------------------------------------------------
+ * implementation of controller operations
+ * ----------------------------------------------------------------- */
+
+SUNControl_ID SUNControlGetID_ARKUserControl(SUNControl C)
+{ return SUNDIALS_CONTROL_H; }
+
+int SUNControlEstimateStep_ARKUserControl(SUNControl C, realtype h,
+                                          realtype dsm, realtype* hnew)
+{
+  /* call user-provided function to compute new step */
+  int retval = SC_HADAPT(C)(SC_ARKMEM(C)->ycur, SC_ARKMEM(C)->tn, h, SC_HP(C),
```

Comment:
Does this need 
```
  ttmp = (dsm <= ONE) ? ark_mem->tn + ark_mem->h : ark_mem->tn;
  nsttmp = (dsm <= ONE) ? ark_mem->nst+1 : ark_mem->nst;
```
to match what would have been passed before?

Reply:
Yes (the `ttmp` part, at least).  Nice catch!